compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -94,10 +98,43 @@ struct PatInfo<'a, 'tcx> {
 
 impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     fn pattern_cause(&self, ti: &TopInfo<'tcx>, cause_span: Span) -> ObligationCause<'tcx> {
+        // If origin_expr exists, then expected represents the type of origin_expr.
+        // If span also exists, then span == origin_expr.span (although it doesn't need to exist).
+        // In that case, we can peel away references from both and treat them
+        // as the same.
+        let origin_expr_info = if let Some(origin_expr) = ti.origin_expr {
+            let mut cur_expr = origin_expr;
+            let mut count = 0;
+
+            // cur_ty may have more layers of references than cur_expr.
+            // We can only make suggestions about cur_expr, however, so we'll
+            // use that as our condition for stopping.
+            while let ExprKind::AddrOf(.., inner) = &cur_expr.kind {
+                cur_expr = inner;
+                count += 1;
+            }
+
+            let needs_parens = match cur_expr.kind {
```

Comment:
This should be a separate function. It seems like something other diagnostics and clippy also have manually written.

Reply:
I'm not sure how to name this in such a way that it's clear that it's only useful for prefix and postfix operations. For example, you wouldn't want to use this for suggesting `a + b` -> `a + b + c` because it would say to put parentheses around `a + b`. Any suggestions? Right now, I'm going with `expr_needs_parens`.

Reply:
Yea that name seems fine. Just add a doc comment explaining what you just told me